### DISCUSSION TITLE
Does xstate on the server require a paradigm shift away from "traditional" OOP?

### DISCUSSION THREAD

**Discussion by universalhandle** [statelyai/xstate#1828] [Q&A]
I'm fascinated by xstate but haven't used it before and could use some help wrapping my head around how it would be used on the server.

Suppose I wanted to model the server for a card game using xstate, where the client is kind of dumb and the server keeps track of gameplay, score, etc. If I were approaching this from a traditional object-oriented programming approach my model might include classes something like these:

```typescript
class Game {
  private id: number;
  private players: Player[];

  public addPlayer(player: Player): void;
  public removePlayer(player: Player): void;
}

class Player {
  id: number;
  isDealer: boolean;
  name: string;
  score: number;

  // methods irrelevant to this example
}
```

For an someone accustomed to OOP, this is satisfying because related data and behavior are grouped together.

As I played around with implementing something similar in xstate, I ended up with something like:

```typescript
const addPlayer = assign({
  players: (context, event) => [...context.players, event.player],
});

const removePlayer = assign({
  players: (context, event) => context.players.filter(player => player.id !== event.player.id),
});

const allPlayersHaveArrived = (context) => context.players.length === 4;

const gameNotEmpty = (context) => context.players.length > 0;

const machine = Machine(
  {
    id: "game",
    initial: "waitingForPlayers",
    context: {
      players: [],
    },
    states: {
      waitingForPlayers: {
        always: {
          cond: "allPlayersHaveArrived",
          target: "dealing"
        },
        on: {
          PLAYER_JOINED: {
            actions: "addPlayer",
            target: "waitingForPlayers"
          },
          PLAYER_LEFT: {
            actions: "removePlayer",
            cond: "gameNotEmpty",
            target: "waitingForPlayers"
          }
        }
      },
      dealing: {
        on: {
          PLAYER_LEFT: "terminated"
        }
      },
      terminated: {
        type: "final"
      }
    }
  },
  {
    actions: { addPlayer, removePlayer },
    guards: { allPlayersHaveArrived, gameNotEmpty }
  }
);

const service = interpret(machine, { devTools: true });
service.start();
```

A notable difference, coming from my background, is that the logical connection between the data and the behavior isn't as clear. Even if I made `addPlayer` and `removePlayer`methods of some object, the data still lives in context. My guards could ignore `context` and reference a `Game` instance, but... I feel like that might be abusing the library and bring unexpected surprises.

Does embracing xstate's approach mean setting this class-based way of thinking aside? Am I failing to commit to the paradigm shift?

**Comment by davidkpiano** [comment]
It's an interesting question, because classes are the quintessential example of an implicit statechart, and class instances are actors.

You can think of methods as events (and in RPC, they basically are!), but finite state is implicit, usually represented by a lot of defensive programming inside the methods, rather than being a first-class concept (pun not intended).

- **Statecharts** can be thought of as **classes**
- **Actors** (statechart instances, e.g., services) can be thought of as **class invocations**

In fact, in many statechart implementations in other languages, statecharts are represented by a class. Lots of similarities there.

If you want to interact with actual class instances, I think it's completely appropriate to treat classes as opaque actors, and invoke an instance of them within a statechart:

```js
createMachine({
  // ...
  invoke: {
    id: 'instance',
    src: (sendBack, onReceive) => {
      const instance = new SomeClass(/* ... */);

      onReceive(event => {/* call methods on instance */});

      return () => { instance.destroy(); } // or some other cleanup
    }
  },
  on: {
    SOME_EVENT: { actions: forwardTo('instance') }
  }
});
```

Now that I think about it, this can be abstracted, either in user-land or as a helper function in XState 🤔 

---

In summary, using statecharts are actually _closer_ to the OOP paradigm than you might think, since OOP itself was originally conceived as the Actor Model, which XState embraces. It's not a paradigm shift, but there's a couple differences:

- You send events instead of calling methods, but they are similar
- Statecharts _formalize_ what can happen in a class based on states. In normal OOP, this is implicit.

So, statecharts are stricter, more formalized classes. Actors are location-transparent class instances (that you send messages to instead of calling methods).

**Reply by universalhandle** [reply]
Thanks for your thoughtful and quick response. Most of what you said immediately made sense to me, and some of it prompted me to make a note to re-read the "Concepts" and "Guides" sections of the documentation to make sure all the words mean what I think they mean :-)

I will get to that in the next few days, and if I have anything smart to say at the end of that, I'll post it here. Many thanks!

**Reply by universalhandle** [reply]
I have decided to learn how to pedal this bicycle before trying to do stunts. The learning curve is steep enough without trying to develop my own dialect, so to speak (and to mix metaphors). That, and I haven't convinced myself that there is a real advantage to working with the library the way I proposed :-)
